gtk/src/toga_gtk/widgets/canvas.py

```diff
@@ -37,22 +47,46 @@ def set_on_resize(self, handler):
         pass
 
     def set_on_press(self, handler):
-        self.interface.factory.not_implemented("Canvas.set_on_press()")
+        self.native.connect("button-press-event", self.mouse_down)
```

Comment:
Good question -- they are additive, yeah. The following example

```python
import gi
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk

def one(a, b):
    print("one")
def two(a, b):
    print("two")

w = Gtk.Window()
w.connect('button-press-event', one)
w.connect('button-press-event', two)
w.show()
Gtk.main()
```

prints, on a single button click (presumably repeated b/c it's not checking for press vs. release),

```
one
two
one
two
```

In order to "overwrite" the existing signal, we have to call `disconnect` on the int returned by the original call to `connect`. I presume you ask because that's the desired behavior so I'll look at updating that shortly.

Reply:
The other approach is to unconditionally hook up the GTK event as part of widget construction, and only invoke the Toga handler if one has been registered. You'll see there's a couple of event handlers registered in the `__init__` method; if you add to that list, Toga will always get a single event, and then leave it up to the *Toga* code to determine if there's a handler to invoke.

Reply:
Yep, that'd be the better way to do it (should've spent more time looking at the other implementations which do exactly that :upside_down_face:)